src/org/exist/storage/journal/JournalReader.java

```diff
@@ -218,23 +235,51 @@ Loggable readEntry() throws LogException {
             }
             loggable.setLsn(lsn);
 
-            if (size + LOG_ENTRY_BACK_LINK_LEN > payload.capacity()) {
+            final int remainingEntryBytes = size + LOG_ENTRY_BACK_LINK_LEN + LOG_ENTRY_CHECKSUM_LEN;
+
+            if (remainingEntryBytes > payload.capacity()) {
                 // resize the payload buffer
-                payload = ByteBuffer.allocate(size + LOG_ENTRY_BACK_LINK_LEN);
+                payload = ByteBuffer.allocateDirect(remainingEntryBytes);
             }
-            payload.clear().limit(size + LOG_ENTRY_BACK_LINK_LEN);
+            payload.clear().limit(remainingEntryBytes);
             read = fc.read(payload);
-            if (read < size + LOG_ENTRY_BACK_LINK_LEN) {
+            if (read < remainingEntryBytes) {
                 throw new LogException("Incomplete log entry found!");
             }
             payload.flip();
+
+            // read entry data
             loggable.read(payload);
-            final short prevLink = payload.getShort();
-            if (prevLink != size + LOG_ENTRY_HEADER_LEN) {
-                LOG.error("Bad pointer to previous: prevLink = " + prevLink + "; size = " + size +
+
+            // read entry backLink
+            final short backLink = payload.getShort();
+            if (backLink != size + LOG_ENTRY_HEADER_LEN) {
+                LOG.error("Bad pointer to previous: backLink = " + backLink + "; size = " + size +
                         "; transactId = " + transactId);
                 throw new LogException("Bad pointer to previous in entry: " + loggable.dump());
             }
+
+            // update the checksum for the entry data and backLink
+            if (payload.hasArray()) {
```

Comment:
looks like almost identical code of line 211 ?

Reply:
I could abstract it, but I would need access to several local vars, as it was small I just reproduced it.